### PR TITLE
Middleware resolvers

### PR DESCRIPTION
**Pull request by afallou**
There are action that we may want to perform for all (or almost all) resolvers in an app.
Currently the only way to do that is to add this action to every resolver. Middleware resolvers allow us to instead centrally define both those actions and the resolvers to which they shouldn't apply.
- Create the MiddlewareResolver class and its interface
- Execute middleware resolvers in any resolver pipeline, before the actual resolver logic.

### CODE REVIEW

File: src/resolvers/middleware.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        if (!this.shouldIgnore(resolverKeys)) {
+            return this.execute(obj, args, context, info);
+        }
+        return [obj, args, context, info];
+    }
+
+    /* If any of the keys in the resolver pipeline should be ignored, ignore the whole pipeline,
+     * making the middleware the no-op for this request.
+     *
+     */
```
Comment (afallou) [inline src/resolvers/middleware.js]: An example is authorization: we may want to restrict write operations to a given role by default, except for some specific ones.

File: src/resolvers/__tests__/pipeline.test.js [hunk 2]
```diff
@@ -100,4 +106,49 @@ describe('Resolver pipeline', () => {
...
+        });
+        setupResolverMiddlewares([tripleFooMiddleware]);
+
+        // The middleware is ignored, because 'identity' is an excluded resolver
+        expect(
+            await getResolverPipeline('foo', 'identity', 'plusOne')(input),
+        ).toEqual(3);
+    });
+
+    it('run middlewares sequentially in order', async() => {
```
Comment (afallou) [inline src/resolvers/__tests__/pipeline.test.js]: This test succeeds in isolation but fails when running with the others. The issue is improper cleanup of the `graphql.resolverMiddlewares` binding, but I don't fully understand it yet.

**Comment by jessemyers** [comment]
@afallou You may want to look a bit at what, if anything, the core GraphQL layer provides.

I don't mind this approach so much as to block it, but I worry about coupling. We're layering a new middleware abstraction on top of our own resolver pipeline abstraction, which means we're going further and further in a direction that doesn't play nice with any other approach to writing resolvers.

At the time I wrote the initial `createResolver`/`createStrictResolver`/`ResolverPipeline` code, writing our own building blocks was justified as there wasn't much out there. I _suspect_ that things have advanced since then, but I have no proof.

**Comment by afallou** [comment]
@jessemyers I looked through the [changelog](https://github.com/graphql/graphql-js/releases) of `graphql-js` and didn't see anything pointing in this direction. There's actually a [feature request](https://github.com/graphql/graphql-js/issues/1516) for resolver middlewares, filed in September.

**Comment by afallou** [comment]
It turns out that the state cleanup issue I'm seeing in one test comes from the fact that nodule-config's `clearBinding` is broken for nested bindings (e.g. `foo.bar`).

**Pull request closed without merging**